src/plugins/media/media.js

```diff
@@ -0,0 +1,172 @@
+/**
+ * Media Plugin
+ *
+ * The media plugin is a combination of a preeinit and prestepleave plugin. 
+ * It is executed before impress:init and before impress:stepleave.
+ * 
+ * It will do the following things:
+ * 
+ *   - Parse special shortcuts to full html5 audio or video nodes:
+ *     e.g. <div class="media" data-media-source="…" data-media-type="audio/…"></div>
+ *     or   <span class="media" data-media-source="…" data-media-type="video/…"></span>
+ *     
+ *     All type of html tags are allowed. Currently only a single source is supported.
+ *     All nodes will have the control attribute added, except the 'data-media-controls="false"'
+ *     attribute is added.
+ *     
+ *         <div class="media" data-media-source="…" data-media-type="…/…" data-media-controls="false"></div>
+ * 
+ *  - Add a special class when playing and pausing media (removing them when ending)
+ *
+ *  - Autostart videos when entering a step, if the attribute data-media-playonenter is set and not "false".
+ *
+ *  - Pause all video and audio an the active step when leaving it. This can be disabled for 
+ *    indivudual media nodes by adding the data-media-pauseonleave="false" attribute. This attribute
+ *    is inherited from the shortcut described above.
+ *    
+ *    Examples: 
+ *
+ *      <audio contols data-media-pauseonleave="false">
```

Comment:
I very seldom use automatically starting media, so I am not the right person to decide which way is the most convenient for the user.

My most often use case is that:
- Enter a step
- Talk about something on the step
- Start media manually
- Sometimes talk a little bit more
- Leave the step, pausing or stopping all media

I think, it is very seldom that someone leaves a step but wants media to play along. Thus it should not be the default value to let it play, when it was started manually.

> If user re-enters the step, it should start media from the beginning. In other words, the stepleave action is to stop the media, not pause.

Are you sure? Maybe someone (including me) likes to proceed at the very same point with the playback. This is made impossible by stopping media instead of pausing it. Aside from that there is no official `.stop()` method for html5 media as far as I know, we would have to pause it anyhow and set the current position to 0.

I would vote for:
- data-media-autoplay: default false
- data-media-autopause: default true
- data-media-autostop: default true if autopause is true and autoplay is true otherwise false

We would have to define which of the two attributes autopause and autostop takes precedence over the other if opposing values were set. 
`data-media-autoplay="true" data-media-autopause="true" data-media-autostop="false"` would result in pausing media on stepleave and resuming on re-enter, according to my description above.
But what do we do on `data-media-autoplay="false"  data-media-autopause="true" data-media-autostop="true"`? Do later values overwrite earlier defined ones?

Of course one could also think of a different solution:
- data-media-autoplay: default false
- data-media-autostop: default copy from autoplay
- data-media-autopause: default true if autoplay is false and autostop is false otherwise false

Regardless of how it is done, it should be heavily documented, because it is complicated to understand if we make a difference between stopping and pausing, and which values must be set to get the desired behavious.

Reply:
Just realized that there would be no way to stop/pause a manually started media without returning to the containing step. I think this is highly inconvenient for users so should never be the default but require an explicit setting by the user.

Reply:
Ok, good suggestion to have both `autostop` and `autopause`.

One answer is easy for me to give: There is a general rule for plugins, that by default they should not do anything. So in this case, if user adds a video or audio tag to a step, and don't use any `data-` attributes, then user must manually start and stop the media. This is not negotiable. Think about it this way: user may not know of the existence of the media plugin, and we must not do anything surprising when it is not explicitly invoked via `data-media-` attributes.

Beyond that, I think I agree with the most common use case: If `data-media-autoplay="true"` this should imply that `data-media-autostop="true"` too. This way the media stops when leaving the step and starts from the beginning if you ever come back. (For example, a presentation might rotate around, in which case it is likely that you return to same slide again.)

*What should be the relation between autostop and autopause?*

IMO if both autostop and autopause are true, the logical result of this is "do both". In practice this is the same as autostop alone. You pause the media, then pause the media and return to beginning. (Or, pause and return to beginning, then pause.)

Again, doing this is my preference simply because it is what you would logically expect. It's an non-default edge case anyway, and you can get the behavior you want by explicitly setting true and false to get what you want.

***

So, as an example of how to achieve your preferred use case, you would do:

    <audio data-media-autoplay="false" data-media-autostop="false" data-media-autopause="true">

or simply:

    <audio data-media-autopause="true">

> Regardless of how it is done, it should be heavily documented, because it is complicated to understand if we make a difference between stopping and pausing, and which values must be set to get the desired behaviour.

Currently plugins are documented with a README.md file in the same directory with the rest of the plugin.

Reply:
Btw, you don't need to implement any code for this, but in documentation you could add an example for how to automatically move to next step when the media reached its end:

    <audio onended="impress().next()">

This is a nice way of creating narrated presentations where presentation moves forward after the media clip finished playing. The above allows for tighter synchronization than for example setting step durations.

Reply:
Okay, you are right with non-surprising default values. But then I would like to have that attributes at the root element and individual steps as well, so that the user is able to overwrite this default behaviour without being forced to add these attributes to each single media.